ace/ace_test.go

```diff
@@ -197,34 +198,125 @@ func Benchmark_Ace(b *testing.B) {
 	engine.AddFunc("isAdmin", func(user string) bool {
 		return user == admin
 	})
-	var buf bytes.Buffer
-	var err error
+	require.NoError(b, engine.Load())
 
 	b.Run("simple", func(bb *testing.B) {
 		bb.ReportAllocs()
 		bb.ResetTimer()
 		for i := 0; i < bb.N; i++ {
-			buf.Reset()
-			err = engine.Render(&buf, "simple", map[string]interface{}{
+			var buf bytes.Buffer
+			//nolint:gosec,errcheck // Return value not needed for benchmark
+			_ = engine.Render(&buf, "simple", map[string]interface{}{
 				"Title": "Hello, World!",
 			})
 		}
-
-		require.NoError(b, err)
-		require.Equal(b, expectSimple, trim(buf.String()))
 	})
 
 	b.Run("extended", func(bb *testing.B) {
 		bb.ReportAllocs()
 		bb.ResetTimer()
 		for i := 0; i < bb.N; i++ {
-			buf.Reset()
-			err = engine.Render(&buf, "extended", map[string]interface{}{
+			var buf bytes.Buffer
+			//nolint:gosec,errcheck // Return value not needed for benchmark
+			_ = engine.Render(&buf, "extended", map[string]interface{}{
 				"User": admin,
 			}, "layouts/main")
 		}
+	})
+
+	b.Run("simple_asserted", func(bb *testing.B) {
+		bb.ReportAllocs()
+		bb.ResetTimer()
+		for i := 0; i < bb.N; i++ {
+			var buf bytes.Buffer
+			err := engine.Render(&buf, "simple", map[string]interface{}{
+				"Title": "Hello, World!",
+			})
 
-		require.NoError(b, err)
-		require.Equal(b, expectExtended, trim(buf.String()))
+			require.NoError(b, err)
```

Comment:
have you found that there is a difference between asserting within the bench loop and outside? I've always seen the assert outside the loop.

Reply:
Asserting outside the loop only checks the LAST value from the loop. Asserting inside the loop exposed that values were changing between loops since the template were not finished rendering.

This was fixed by adding the mutex in `Render()`.